Source/Core/Core/HW/EXI/BBA/XLINK_KAI_BBA.cpp

```diff
@@ -0,0 +1,281 @@
+}
+
+bool CEXIETHERNET::XLinkNetworkInterface::IsActivated()
+{
+  return m_sf_socket.getLocalPort() > 0;
+}
+
+bool CEXIETHERNET::XLinkNetworkInterface::SendFrame(const u8* frame, u32 size)
+{
+  // Overwrite buffer and add header
+  memmove(m_out_frame, "e;e;", 4);
+  memmove(m_out_frame + 4, frame, size);
+  size += 4;
+
+  // Only uncomment for debugging, the performance hit is too big otherwise
+  // INFO_LOG(SP1, "SendFrame %x\n%s", size, ArrayToString(m_out_frame, size, 0x10).c_str());
+
+  if (m_sf_socket.send(m_out_frame, size, m_sf_recipient_ip, m_dest_port) != sf::Socket::Done)
+  {
+    ERROR_LOG(SP1, "SendFrame(): expected to write %d bytes, but failed, errno %d", size, errno);
+    return false;
+  }
+  else
+  {
+    m_eth_ref->SendComplete();
+    return true;
+  }
+}
+
+void CEXIETHERNET::XLinkNetworkInterface::ReadThreadHandler(
+    CEXIETHERNET::XLinkNetworkInterface* self)
+{
+  sf::IpAddress sender;
+  unsigned short port;
+
+  while (!self->readThreadShutdown.IsSet())
+  {
+    if (!self->IsActivated())
+      break;
+
+    // XLink supports jumboframes but Gamecube BBA does not. We need to support jumbo frames
+    // *here* because XLink *could* send one
+    std::size_t bytes_read = 0;
+    if (self->m_sf_socket.receive(self->m_in_frame, 9004, bytes_read, sender, port) !=
+        sf::Socket::Done)
+    {
+      ERROR_LOG(SP1, "Failed to read from BBA, err=%lu", bytes_read);
+    }
+
+    // Did we get an ethernet frame? Copy the first 4 bytes to check the header
+    char tempCheck[4];
+    memmove(tempCheck, self->m_in_frame, 4);
+
+    // Check for e;e; header; this indicates the received data is an ethernet frame
+    // e;e;raw_ethernet_frame_data
+    if (tempCheck[2] == 'e' && tempCheck[3] == ';')
+    {
+      // Is the frame larger than BBA_RECV_SIZE?
+      if ((bytes_read - 4) < BBA_RECV_SIZE)
+      {
+        DEBUG_LOG(SP1, "Received ethernet frame");  // TODO remove this logging for performance?
+
+        // Copy payload into BBA buffer as an ethernet frame
+        memmove(self->m_eth_ref->mRecvBuffer.get(), self->m_in_frame + 4, bytes_read - 4);
+
+        if (bytes_read < 1)
+        {
+          ERROR_LOG(SP1, "Failed to read from BBA, err=%lu", bytes_read - 4);
+        }
+        else if (self->readEnabled.IsSet())
+        {
+          // Only uncomment for debugging, the performance hit is too big otherwise
+          // DEBUG_LOG(SP1, "Read data: %s", ArrayToString(self->m_eth_ref->mRecvBuffer.get(),
+          // (u32)(bytes_read - 4), 0x10).c_str());
+          self->m_eth_ref->mRecvBufferLength = (u32)(bytes_read - 4);
+          self->m_eth_ref->RecvHandlePacket();
+        }
+      }
+    }
+    // Otherwise we recieved control data or junk
+    else
+    {
+      char recv_str[9004] = {0x00};
+      memmove(recv_str, self->m_in_frame, bytes_read);
+      INFO_LOG(SP1, "Received XLink Kai control data: %s", recv_str);
+      std::string controlMsg = std::string(recv_str);
```

Comment:
May want to use a std::string to begin with so you can initialize the data in place like:

```cpp
std::string recv_str(self->m_in_frame, self->m_in_frame + bytes_read);
```

instead of using both the stack and then converting the string over later.